addons/binding/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/internal/handler/SmartHomeHandler.java

```diff
@@ -105,15 +120,64 @@ Connection createConnection(TPLinkSmartHomeConfiguration config) {
         return new Connection(config.ipAddress);
     }
 
-    private DeviceState refreshCache() {
+    private Optional<DeviceState> refreshCache() {
         try {
+            updateIpAddress();
             DeviceState deviceState = new DeviceState(connection.sendCommand(smartHomeDevice.getUpdateCommand()));
-            updateStatus(ThingStatus.ONLINE);
+            updateDeviceId(deviceState.getSysinfo().getDeviceId());
+            if (getThing().getStatus() != ThingStatus.ONLINE) {
+                updateStatus(ThingStatus.ONLINE);
+            }
 
-            return deviceState;
+            return Optional.of(deviceState);
         } catch (IOException e) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
-            return null;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            return Optional.empty();
+        } catch (RuntimeException e) {
```

Comment:
Should you also log the strack trace an NPE for example tends to have message `null`